File: sdk-typescript/packages/chapters/src/chapters.ts
/**
 * WAVE SDK - Chapters API
 *
 * Manage video chapters and smart chapter generation.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginatedResponse,
} from '@wave-av/core';
import type { Chapter, ChapterSet, GenerateChaptersRequest, CreateChapterSetRequest, CreateChapterRequest, UpdateChapterRequest, UpdateChapterSetRequest, ListChapterSetsParams } from './chapters-types';
export type { ChapterStatus, Chapter, ChapterSet, GenerateChaptersRequest, CreateChapterSetRequest, CreateChapterRequest, UpdateChapterRequest, UpdateChapterSetRequest, ListChapterSetsParams } from './chapters-types';

// Types

/**
 * Chapter status
 */

/**
 * Chapter
 */

/**
 * Chapter set (collection of chapters for a media)
 */

/**
 * Generate chapters request
 */

/**
 * Create chapter set request
 */

/**
 * Create chapter request
 */

/**
 * Update chapter request
 */

/**
 * Update chapter set request
 */

/**
 * List chapter sets params
 */

// Chapters API

/**
 * Chapters API client
 */
export class ChaptersAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/chapters';

  constructor(client: WaveClient) {
    this.client = client;
  }

  // Chapter Sets

  /**
   * Generate chapters using AI
   *
   * Requires: chapters:generate permission
   */
  async generate(request: GenerateChaptersRequest): Promise<ChapterSet> {
    return this.client.post<ChapterSet>(`${this.basePath}/generate`, request);
  }

  /**
   * Create a chapter set manually
   *
   * Requires: chapters:create permission
   */
  async createSet(request: CreateChapterSetRequest): Promise<ChapterSet> {
    return this.client.post<ChapterSet>(this.basePath, request);
  }

  /**
   * Get a chapter set by ID
   *
   * Requires: chapters:read permission
   */
  async getSet(setId: string): Promise<ChapterSet> {
    return this.client.get<ChapterSet>(`${this.basePath}/${setId}`);
  }

  /**
   * Update a chapter set
   *
   * Requires: chapters:update permission
   */
  async updateSet(setId: string, request: UpdateChapterSetRequest): Promise<ChapterSet> {
    return this.client.patch<ChapterSet>(`${this.basePath}/${setId}`, request);
  }

  /**
   * Remove a chapter set
   *
   * Requires: chapters:remove permission (canDelete verified server-side)
   */
  async removeSet(setId: string): Promise<void> {
    // canDelete permission enforced by API
    await this.client.delete(`${this.basePath}/${setId}`);
  }

  /**
   * List chapter sets
   *
   * Requires: chapters:read permission
   */
  async listSets(params?: ListChapterSetsParams): Promise<PaginatedResponse<ChapterSet>> {
    return this.client.get<PaginatedResponse<ChapterSet>>(this.basePath, { params: params as Record<string, string | number | boolean | undefined> });
  }

  /**
   * Get the default chapter set for a media
   *
   * Requires: chapters:read permission
   */
  async getDefaultSet(
    mediaId: string,
    mediaType: 'video' | 'audio' | 'stream'
  ): Promise<ChapterSet | null> {
    try {
      return await this.client.get<ChapterSet>(`${this.basePath}/default`, {
        params: { media_id: mediaId, media_type: mediaType } as unknown as Record<string, string | number | boolean | undefined>,
      });
    } catch (error) {
      if ((error as { statusCode?: number }).statusCode === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Duplicate a chapter set
   *
   * Requires: chapters:create permission
   */
  async duplicateSet(setId: string, name?: string): Promise<ChapterSet> {
    return this.client.post<ChapterSet>(`${this.basePath}/${setId}/duplicate`, { name });
  }

  // Individual Chapters

  /**
   * Add a chapter to a set
   *
   * Requires: chapters:update permission
   */
  async addChapter(setId: string, chapter: CreateChapterRequest): Promise<Chapter> {
    return this.client.post<Chapter>(`${this.basePath}/${setId}/chapters`, chapter);
  }

  /**
   * Get a chapter by ID
   *
   * Requires: chapters:read permission
   */
  async getChapter(setId: string, chapterId: string): Promise<Chapter> {
    return this.client.get<Chapter>(`${this.basePath}/${setId}/chapters/${chapterId}`);
  }

  /**
   * Update a chapter
   *
   * Requires: chapters:update permission
   */
  async updateChapter(
    setId: string,
    chapterId: string,
    request: UpdateChapterRequest
  ): Promise<Chapter> {
    return this.client.patch<Chapter>(
      `${this.basePath}/${setId}/chapters/${chapterId}`,
      request
    );
  }

  /**
   * Remove a chapter
   *
   * Requires: chapters:update permission (server-side RBAC enforced)
   */
  async removeChapter(setId: string, chapterId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${setId}/chapters/${chapterId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Reorder chapters
   *
   * Requires: chapters:update permission
   */
  async reorderChapters(
    setId: string,
    chapterIds: string[]
  ): Promise<ChapterSet> {
    return this.client.post<ChapterSet>(
      `${this.basePath}/${setId}/chapters/reorder`,
      { chapter_ids: chapterIds }
    );
  }

  /**
   * Bulk update chapters
   *
   * Requires: chapters:update permission
   */
  async bulkUpdateChapters(
    setId: string,
    updates: Array<{ id: string } & Partial<UpdateChapterRequest>>
  ): Promise<{ updated: number }> {
    return this.client.post(`${this.basePath}/${setId}/chapters/bulk`, { updates });
  }

  // Thumbnails

  /**
   * Generate thumbnail for a chapter
   *
   * Requires: chapters:update permission
   */
  async generateThumbnail(
    setId: string,
    chapterId: string,
    options?: { time?: number }
  ): Promise<Chapter> {
    return this.client.post<Chapter>(
      `${this.basePath}/${setId}/chapters/${chapterId}/thumbnail`,
      options
    );
  }

  /**
   * Generate thumbnails for all chapters in a set
   *
   * Requires: chapters:update permission
   */
  async generateAllThumbnails(setId: string): Promise<{ generated: number }> {
    return this.client.post(`${this.basePath}/${setId}/thumbnails`);
  }

  // Export

  /**
   * Export chapters in various formats
   *
   * Requires: chapters:read permission
   */
  async exportChapters(
    setId: string,
    format: 'json' | 'youtube' | 'webvtt' | 'ffmpeg'
  ): Promise<{ content: string; format: string }> {
    return this.client.get(`${this.basePath}/${setId}/export`, {
      params: { format } as unknown as Record<string, string | number | boolean | undefined>,
    });
  }

  /**
   * Import chapters from a format
   *
   * Requires: chapters:create permission
   */
  async importChapters(
    mediaId: string,
    mediaType: 'video' | 'audio' | 'stream',
    format: 'json' | 'youtube' | 'webvtt' | 'ffmpeg',
    content: string,
    options?: { name?: string; set_as_default?: boolean }
  ): Promise<ChapterSet> {
    return this.client.post<ChapterSet>(`${this.basePath}/import`, {
      media_id: mediaId,
      media_type: mediaType,
      format,
      content,
      ...options,
    });
  }

  // Utilities

  /**
   * Wait for chapter generation to complete
   */
  async waitForReady(
    setId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (set: ChapterSet) => void;
    }
  ): Promise<ChapterSet> {
    const pollInterval = options?.pollInterval || 2000;
    const timeout = options?.timeout || 600000; // 10 minutes
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const set = await this.getSet(setId);

      if (options?.onProgress) {
        options.onProgress(set);
      }

      if (set.status === 'ready') {
        return set;
      }

      if (set.status === 'failed') {
        throw new Error(`Chapter generation failed: ${set.error || 'Unknown error'}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Chapter generation timed out after ${timeout}ms`);
  }

  /**
   * Get chapter at a specific time
   *
   * Requires: chapters:read permission
   */
  async getChapterAtTime(setId: string, time: number): Promise<Chapter | null> {
    try {
      return await this.client.get<Chapter>(`${this.basePath}/${setId}/at`, {
        params: { time } as unknown as Record<string, string | number | boolean | undefined>,
      });
    } catch (error) {
      if ((error as { statusCode?: number }).statusCode === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Merge chapters
   *
   * Requires: chapters:update permission
   */
  async mergeChapters(
    setId: string,
    chapterIds: string[],
    options?: { title?: string; description?: string }
  ): Promise<Chapter> {
    return this.client.post<Chapter>(`${this.basePath}/${setId}/chapters/merge`, {
      chapter_ids: chapterIds,
      ...options,
    });
  }

  /**
   * Split a chapter at a specific time
   *
   * Requires: chapters:update permission
   */
  async splitChapter(
    setId: string,
    chapterId: string,
    splitTime: number,
    options?: { first_title?: string; second_title?: string }
  ): Promise<{ first: Chapter; second: Chapter }> {
    return this.client.post(`${this.basePath}/${setId}/chapters/${chapterId}/split`, {
      split_time: splitTime,
      ...options,
    });
  }
}

/**
 * Create a Chapters API instance
 */
export function createChaptersAPI(client: WaveClient): ChaptersAPI {
  return new ChaptersAPI(client);
}

File: sdk-python/wave_av_sdk/chapters.py
"""WAVE SDK - Chapters API. Auto-generate and manage video chapters."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Chapter(BaseModel):
    id: str; title: str; start_time: float; end_time: float; thumbnail_url: str | None = None; description: str | None = None

class ChapterSet(BaseModel):
    id: str; asset_id: str; status: str; chapters: list[Chapter] | None = None; created_at: str; updated_at: str

class ChaptersAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/chapters"
    def generate(self, asset_id: str, **kwargs: Any) -> ChapterSet: return ChapterSet(**self._client.post(self._base, json={"asset_id": asset_id, **kwargs}))
    def create_set(self, asset_id: str, name: str = "Custom", **kwargs: Any) -> ChapterSet: return ChapterSet(**self._client.post(f"{self._base}/sets", json={"asset_id": asset_id, "name": name, **kwargs}))
    def get_set(self, set_id: str) -> ChapterSet: return ChapterSet(**self._client.get(f"{self._base}/{set_id}"))
    def list_sets(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def update_set(self, set_id: str, **kwargs: Any) -> ChapterSet: return ChapterSet(**self._client.patch(f"{self._base}/{set_id}", json=kwargs))
    def remove_set(self, set_id: str) -> None: self._client.delete(f"{self._base}/{set_id}")
    def duplicate_set(self, set_id: str) -> ChapterSet: return ChapterSet(**self._client.post(f"{self._base}/{set_id}/duplicate"))
    def get_default_set(self, asset_id: str) -> ChapterSet: return ChapterSet(**self._client.get(f"{self._base}/default/{asset_id}"))
    def add_chapter(self, set_id: str, title: str, start_time: float, end_time: float, **kwargs: Any) -> Chapter: return Chapter(**self._client.post(f"{self._base}/{set_id}/chapters", json={"title": title, "start_time": start_time, "end_time": end_time, **kwargs}))
    def get_chapter(self, set_id: str, chapter_id: str) -> Chapter: return Chapter(**self._client.get(f"{self._base}/{set_id}/chapters/{chapter_id}"))
    def update_chapter(self, set_id: str, chapter_id: str, **kwargs: Any) -> Chapter: return Chapter(**self._client.patch(f"{self._base}/{set_id}/chapters/{chapter_id}", json=kwargs))
    def remove_chapter(self, set_id: str, chapter_id: str) -> None: self._client.delete(f"{self._base}/{set_id}/chapters/{chapter_id}")
    def get_chapter_at_time(self, set_id: str, time_seconds: float) -> Chapter: return Chapter(**self._client.get(f"{self._base}/{set_id}/at-time", params={"time": time_seconds}))
    def reorder_chapters(self, set_id: str, chapter_ids: list[str]) -> list[Chapter]: return [Chapter(**c) for c in self._client.put(f"{self._base}/{set_id}/reorder", json={"chapter_ids": chapter_ids})]
    def split_chapter(self, set_id: str, chapter_id: str, split_time: float) -> list[Chapter]: return [Chapter(**c) for c in self._client.post(f"{self._base}/{set_id}/chapters/{chapter_id}/split", json={"split_time": split_time})]
    def merge_chapters(self, set_id: str, chapter_ids: list[str]) -> Chapter: return Chapter(**self._client.post(f"{self._base}/{set_id}/merge", json={"chapter_ids": chapter_ids}))
    def bulk_update_chapters(self, set_id: str, chapters: list[dict]) -> list[Chapter]: return [Chapter(**c) for c in self._client.put(f"{self._base}/{set_id}/chapters", json={"chapters": chapters})]
    def generate_thumbnail(self, set_id: str, chapter_id: str) -> Chapter: return Chapter(**self._client.post(f"{self._base}/{set_id}/chapters/{chapter_id}/thumbnail"))
    def generate_all_thumbnails(self, set_id: str) -> ChapterSet: return ChapterSet(**self._client.post(f"{self._base}/{set_id}/thumbnails"))
    def export_chapters(self, set_id: str, format: str = "json") -> dict: return self._client.get(f"{self._base}/{set_id}/export", params={"format": format})
    def import_chapters(self, asset_id: str, format: str, content: str) -> ChapterSet: return ChapterSet(**self._client.post(f"{self._base}/import", json={"asset_id": asset_id, "format": format, "content": content}))
    def wait_for_ready(self, set_id: str, poll_interval: float = 2.0, timeout: float = 300.0) -> ChapterSet:
        start = time.time()
        while time.time() - start < timeout:
            s = self.get_set(set_id)
            if s.status == "ready": return s
            if s.status == "failed": raise RuntimeError("Chapter generation failed")
            time.sleep(poll_interval)
        raise TimeoutError(f"Chapters timed out after {timeout}s")
